File: android/src/main/java/tech/geelab/reactnative/geelabguard/GeelabGuardReceiptMapper.kt
package tech.geelab.reactnative.geelabguard

import com.facebook.react.bridge.Arguments
import com.facebook.react.bridge.WritableMap
import kotlin.io.encoding.Base64
import kotlin.io.encoding.ExperimentalEncodingApi

@OptIn(ExperimentalEncodingApi::class)
internal class GeelabGuardReceiptMapper(
  private val mapFactory: () -> WritableMap = { Arguments.createMap() },
) {
  fun toWritableMap(receipt: GeelabGuardNativeReceipt): WritableMap = mapFactory().apply {
    putNullableString("appId", receipt.appId)
    putNullableString("geeToken", receipt.geeToken)
    putNullableString("geeId", receipt.geeId)
    putNullableString("geeIdTimestamp", receipt.geeIdTimestamp)
    putNullableString("respondedGeeToken", receipt.respondedGeeToken)
    putNullableString(
      "originalResponseBase64",
      receipt.originalResponse?.let { Base64.encode(it.toByteArray(Charsets.UTF_8)) },
    )
  }

  fun toErrorUserInfo(
    nativeCode: Int,
    receipt: GeelabGuardNativeReceipt?,
  ): WritableMap = mapFactory().apply {
    putInt("nativeCode", nativeCode)
    receipt?.let { putMap("receipt", toWritableMap(it)) }
  }

  private fun WritableMap.putNullableString(key: String, value: String?) {
    if (value == null) putNull(key) else putString(key, value)
  }
}

File: ios/GeelabGuardReceiptMapper.swift
import Foundation

struct GeelabGuardReceiptMapper {
  func toDictionary(_ receipt: GeelabGuardNativeReceipt) -> [String: Any] {
    [
      "appId": nullable(receipt.appId),
      "geeToken": nullable(receipt.geeToken),
      "geeId": nullable(receipt.geeId),
      "geeIdTimestamp": nullable(receipt.geeIdTimestamp),
      "respondedGeeToken": nullable(receipt.respondedGeeToken),
      "originalResponseBase64": nullable(
        receipt.originalResponse?.base64EncodedString()
      ),
    ]
  }

  private func nullable(_ value: Any?) -> Any {
    value ?? NSNull()
  }
}
